## Environment 1: LunarLander-v3

In [1]:
import gymnasium as gym

env1 = gym.make("LunarLander-v3", render_mode="human")
observation, info = env1.reset()

for _ in range(500):
    action = env1.action_space.sample()  # Take random action
    observation, reward, terminated, truncated, info = env1.step(action)
    if terminated or truncated:
        observation, info = env1.reset()

env1.close()

In [5]:
env1 = gym.make("LunarLander-v3")
print("LunarLander-v3")
print("Observation space:", env1.observation_space)
print("Action space:", env1.action_space)
env1.close()

LunarLander-v3
Observation space: Box([ -2.5        -2.5       -10.        -10.         -6.2831855 -10.
  -0.         -0.       ], [ 2.5        2.5       10.        10.         6.2831855 10.
  1.         1.       ], (8,), float32)
Action space: Discrete(4)


## LunarLander-v3

- **Observation space**: Box(8,) — an 8-dimensional vector with bounds [-2.5 to 2.5, -2.5 to 2.5, -10 to 10, -10 to 10, -6.28 to 6.28, -10 to 10, 0 to 1, 0 to 1]:
  - Elements 0-1: Lander position (x, y coordinates)
  - Elements 2-3: Lander velocity (x, y components) 
  - Element 4: Lander angle (in radians, ±π)
  - Element 5: Lander angular velocity
  - Elements 6-7: Boolean leg contact sensors (left leg, right leg)
- **Action space**: Discrete(4) — 4 discrete actions: do nothing, fire left engine, fire main engine, fire right engine.
- **Goal/objective**: Land the lunar module softly on the landing pad.
- **Typical episode length**: About 500–1000 steps if not terminated sooner.
- **Challenges**: Balancing thrust, landing at the right speed and angle, not crashing.

---

## Environment 2: CarRacing

In [3]:
env2 = gym.make("CarRacing-v3", render_mode="human")
observation, info = env2.reset(seed=42)

for _ in range(500):
    action = env2.action_space.sample()
    observation, reward, terminated, truncated, info = env2.step(action)
    if terminated or truncated:
        observation, info = env2.reset()

env2.close()

In [7]:
env2 = gym.make("CarRacing-v3")
print("\nCarRacing-v3")
print("Observation space:", env2.observation_space)
print("Action space:", env2.action_space)
env2.close()


CarRacing-v3
Observation space: Box(0, 255, (96, 96, 3), uint8)
Action space: Box([-1.  0.  0.], 1.0, (3,), float32)


## CarRacing-v3

- **Observation space**: Box(0, 255, (96, 96, 3), uint8) — 96×96×3 RGB image with pixel values 0-255, representing the car and track view.
- **Action space**: Box([-1, 0, 0], [1, 1, 1], (3,), float32) — continuous control:
  - Action 0: Steering [-1, 1] (left/right)
  - Action 1: Gas [0, 1] (throttle)
  - Action 2: Brake [0, 1] (braking force)
- **Goal/objective**: Drive as far as possible around the procedurally generated track.
- **Typical episode length**: Up to 1000–2000 frames.
- **Challenges**: Keeping the car on track, handling curves, speed control with continuous actions.